### DISCUSSION TITLE
Python Virtualenvs

### DISCUSSION THREAD

**Discussion by IceS2** [helix-editor/helix#5348] [Q&A] [answered]
Hello folks! Amazing work!
I'd really love to try Helix out instead of my Neovim setup but I'd love to understand if it works with Python Virtualenvs. I've been trying to find an answer to this but didn't

Thanks a lot (=

**Comment by kirawi** [comment, marked as answer]
Aren't virtualenvs related to the LSP instead?

**Reply by pascalkuthe** [reply]
Yeah virtualenvs don't need to/aren't integrated with the editor directly usually. You just enter the virtualemv in a terminal and start your editor afterwards

**Reply by IceS2** [reply]
I agree with you both, but there are many editors that automatically detect the right python interpreter and I couldn't figure out how to do so in Helix.

I'll attach here a commented example of my Neovim configuration for this:


```lua
local function get_python_path(workspace)
  -- Use activated virtualenv.
  if vim.env.VIRTUAL_ENV then
    return path.join(vim.env.VIRTUAL_ENV, 'bin', 'python')
  end

  -- Find and use virtualenv in workspace directory.
  for _, pattern in ipairs({'*', '.*'}) do
    local match = vim.fn.glob(path.join(workspace, pattern, '.python-local'))
    if match ~= '' then
      return path.join(vim.env.PYENV_ROOT, path.dirname(match), 'bin', 'python')
    end
  end

  -- Fallback to system Python.
  return exepath('python3') or exepath('python') or 'python'
end


-- Actually setup the installed languages
local lsp_installer = require("nvim-lsp-installer")
lsp_installer.on_server_ready(function(server)
  local opts = {
    capabilities = require("cmp_nvim_lsp").update_capabilities(vim.lsp.protocol.make_client_capabilities()),
  }

-- Here I'm actually passing to pyright the right pythonPath based on the virtualenv I'm using.
  if server.name == "pyright" then
    opts.on_init = function(client)
        client.config.settings.python.pythonPath = get_python_path(client.config.root_dir)
      end
  end
  server:setup(opts)
end)
```

Since the configuration is on the TOML file I couldn't understand if there was a way to do something similar. I'm not familiar with other editor's configuration to be honest but I am pretty sure there are some relative easy way to automatically select your python interpreter in PyCharm/VSCode

**Reply by kirawi** [reply]
I'm not familiar with Python, but there are two ways you could go about it. One would be writing a wrapper script over the LSP to do that for you. The second would be creating a `.helix/languages.toml` file with [the correct config](https://docs.helix-editor.com/languages.html#language-server-configuration) per each project you have (perhaps automating the creation with an alias). This would be better addressed by the future plugin system.

**Reply by IceS2** [reply]
I see.
For the first one, I could pass an arbitraty script here `language-server = { command = "MY SCRIPT", args = ["MY ARGS"] }`, right?
That seems like a great idea. I'll try it out when I have some time!

Glad to know that worst case I can configure it per project basis! I'm a bit lazy for that but I've already got a wrapper script to create a virtualenv so I could fit that in as well in a worst case scenario.

It might take a few days for me to test it just becasue new years eve but I'll come back with my results!

Thanks a lot and amazing project! I've been a Neovim user for a while and I'm really looking forward to testing Helix in a day-to-day basis (=

**Reply by kirawi** [reply]
Yep!

**Reply by IceS2** [reply]
Hey! How are you folks? I've just tested the idea of having a script but it's not working out... let me share with you my different bits to see if you can help me out (=

1. Script (I'm using a symlink to /usr/local/bin/pyright-ls-helix)
```bash
if [[ $VIRTUAL_ENV ]]; then
  python_path="${VIRTUAL_ENV}/bin/python"
else
  python_path="python"
fi

export PYTHONPATH=python_path

pyright-langserver --stdio
```

2. languages.toml
```toml
[[language]]
name = "python"
language-server = { command = "pyright-ls-helix" }
config = {}
```

3. helix --health python
```
❯ helix --health python
Configured language server: pyright-ls-helix
Binary for language server: /usr/local/bin/pyright-ls-helix
Configured debug adapter: None
Highlight queries: ✓
Textobject queries: ✓
Indent queries: ✓
```

4. Log from Helix
```
3.059 helix_view::editor [ERROR] Failed to initialize the LSP for `source.python` { cannot find binary path }
3.502 helix_view::editor [ERROR] Failed to initialize the LSP for `source.python` { IO Error: Exec format error (os error 8) }
```
(On another topic... For some reason yank to system clipboard didn't work and I had to write the logs here myself)

**Reply by kirawi** [reply]
I found this issue for that: https://stackoverflow.com/questions/27606653/oserror-errno-8-exec-format-error
Alternatively, you could do something like: `language-server = { command = "python", args = "pyright-ls-helix" }`
As for the clipboard: https://github.com/helix-editor/helix/wiki/Troubleshooting#copypaste-fromto-system-clipboard-not-working

**Reply by IceS2** [reply]
Hey Kirawi! Thanks for the reply and so sorry for asking about those errors that weren't related. I thought it could be internal to Helix...

I actually forgot to add the shebang line and now it's actually trying to initialize it but I'm running into another error.

`2023-01-02T15:14:59.703 helix_term::application [ERROR] Language Server: Method not found client/registerCapability`

Tomorrow I'll try to debug a bit more and see if I understand what's going on. My bet is that I need to pass some extra configuration to pyright as it's being passed `config = {}` at the moment.

As for the Clipboard... It was an issue between the chair and the keyboard ;P
(I forgot to select the line before yanking since I'm used to vim)

**Reply by kirawi** [reply]
I think the LSP will probably work despite that error.

**Reply by IceS2** [reply]
I tested it quickly with <Space + a> and it prompted me a message saying that the LSP wasn't working, that's why I think it might be a blocking error

**Reply by kirawi** [reply]
To help diagnose why the LSP may not be working, you could start helix in a higher logging verbosity mode via `hx -v`.

**Reply by IceS2** [reply]
Alright... I've got a curious update... After coming back again today for some more testing... It just works 

Have you tried turn it off and on again? :shrug: 

Thanks a lot for all the support Kirawi! I'll try to test it out a little since it seems amazing and I'll continue keeping an eye on the Issue about a File Tree (=

I'll leave here the configuration that made it work just in case anyone searches for this (=

1. languages.toml
```toml
[[language]]
name = "python"
language-server = { command = "pyright-ls-helix" }
config = {}
```

2. pyright-ls-helix script (Using a symlink just to have it on path `ln -s $HOME/.config/helix/scripts/pyright-ls.sh /usr/local/bin/pyright-ls-helix`)
```bash
#!/usr/bin/bash

if [[ $VIRTUAL_ENV ]]; then
  python_path="${VIRTUAL_ENV}/bin/python"
else
  python_path="python"
fi

export PYTHONPATH=python_path

pyright-langserver --stdio
```

**Comment by Aceticia** [comment]
Just in case someone needs this later, I made a small change to make it also work for conda:
```zsh
if [[ $VIRTUAL_ENV ]]; then
  python_path="${VIRTUAL_ENV}/bin/python"
elif [[ $CONDA_PREFIX ]] ; then
  python_path="${CONDA_PREFIX}/bin/python"
else
  python_path="python"
fi

export PYTHONPATH=python_path
```
